File: apps/api/landsignal/scoring/financial.py
from __future__ import annotations


def clamp(n: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, n))


def price_per_acre(price_usd: float | None, acreage: float | None) -> float | None:
    if price_usd is None or acreage is None or acreage <= 0:
        return None
    return price_usd / acreage


def asking_discount_pct(asking: float | None, estimated: float | None) -> float | None:
    if asking is None or estimated is None or estimated <= 0 or asking < 0:
        return None
    return ((asking - estimated) / estimated) * 100.0


def margin_of_safety(purchase: float, base_value: float) -> float:
    if base_value <= 0:
        return 0.0
    return (base_value - purchase) / base_value


def noi_from_rent(
    gross_rent: float,
    vacancy_rate: float,
    opex: float,
    taxes: float,
    insurance: float,
    management: float,
) -> float:
    egi = gross_rent * (1 - vacancy_rate)
    return egi - opex - taxes - insurance - management


def cap_rate(noi: float, value: float) -> float | None:
    if value <= 0:
        return None
    return noi / value


def cash_on_cash(annual_cf: float, equity: float) -> float | None:
    if equity <= 0:
        return None
    return annual_cf / equity


def npv(rate: float, cash_flows: list[float]) -> float:
    total = 0.0
    for t, cf in enumerate(cash_flows):
        total += cf / ((1 + rate) ** t)
    return total


def irr(
    cash_flows: list[float],
    lo: float = -0.99,
    hi: float = 10.0,
    tol: float = 1e-6,
    max_iter: int = 200,
) -> float | None:
    if len(cash_flows) < 2:
        return None
    npv_lo = npv(lo, cash_flows)
    npv_hi = npv(hi, cash_flows)
    if npv_lo * npv_hi > 0:
        return None
    for _ in range(max_iter):
        mid = (lo + hi) / 2
        npv_mid = npv(mid, cash_flows)
        if abs(npv_mid) < tol:
            return mid
        if npv_lo * npv_mid <= 0:
            hi = mid
            npv_hi = npv_mid
        else:
            lo = mid
            npv_lo = npv_mid
    return (lo + hi) / 2


def breakeven_land_value(noi: float, target_cap: float) -> float | None:
    if target_cap <= 0:
        return None
    return noi / target_cap


def farmland_scenario(
    cash_rent_per_acre: float,
    acres: float,
    vacancy_rate: float,
    opex_per_acre: float,
    taxes: float,
    insurance: float,
    management: float,
    purchase_price: float,
    hold_years: int,
    exit_cap_rate: float,
    annual_appreciation: float,
    discount_rate: float,
) -> dict:
    gross = cash_rent_per_acre * acres
    noi = noi_from_rent(
        gross,
        vacancy_rate,
        opex_per_acre * acres,
        taxes,
        insurance,
        management,
    )
    flows = [-purchase_price] + [noi] * hold_years
    exit_value = purchase_price * ((1 + annual_appreciation) ** hold_years)
    flows[-1] += exit_value
    return {
        "gross_rent": gross,
        "noi": noi,
        "cap_rate": cap_rate(noi, purchase_price),
        "cash_on_cash": cash_on_cash(noi, purchase_price),
        "irr": irr(flows),
        "npv": npv(discount_rate, flows),
        "breakeven_land_value": breakeven_land_value(noi, exit_cap_rate),
    }

File: packages/scoring/src/financial.ts
/**
 * Pure financial helpers for land underwriting.
 * All monetary units USD unless noted. Rates are decimals (0.08 = 8%).
 */

export function clamp(n: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, n));
}

export function pricePerAcre(
  priceUsd: number | null | undefined,
  acreage: number | null | undefined,
): number | null {
  if (priceUsd == null || acreage == null || acreage <= 0) return null;
  return priceUsd / acreage;
}

export function askingDiscountPct(
  askingUsd: number | null | undefined,
  estimatedValueUsd: number | null | undefined,
): number | null {
  if (
    askingUsd == null ||
    estimatedValueUsd == null ||
    estimatedValueUsd <= 0 ||
    askingUsd < 0
  ) {
    return null;
  }
  // Negative => asking below estimate (discount to buyer)
  return ((askingUsd - estimatedValueUsd) / estimatedValueUsd) * 100;
}

export function marginOfSafety(
  purchaseUsd: number,
  baseValueUsd: number,
): number {
  if (baseValueUsd <= 0) return 0;
  return (baseValueUsd - purchaseUsd) / baseValueUsd;
}

export function noiFromRent(
  grossRent: number,
  vacancyRate: number,
  opex: number,
  taxes: number,
  insurance: number,
  management: number,
): number {
  const egI = grossRent * (1 - vacancyRate);
  return egI - opex - taxes - insurance - management;
}

export function capRate(noi: number, value: number): number | null {
  if (value <= 0) return null;
  return noi / value;
}

export function cashOnCash(annualCashFlow: number, equityIn: number): number | null {
  if (equityIn <= 0) return null;
  return annualCashFlow / equityIn;
}

/** NPV of cash flows; cashFlows[0] is t=0 (usually negative purchase). */
export function npv(discountRate: number, cashFlows: number[]): number {
  return cashFlows.reduce(
    (acc, cf, t) => acc + cf / Math.pow(1 + discountRate, t),
    0,
  );
}

/**
 * IRR via bisection. Returns null if no bracketed root.
 */
export function irr(
  cashFlows: number[],
  guessLow = -0.99,
  guessHigh = 10,
  tol = 1e-6,
  maxIter = 200,
): number | null {
  if (cashFlows.length < 2) return null;
  let lo = guessLow;
  let hi = guessHigh;
  let npvLo = npv(lo, cashFlows);
  let npvHi = npv(hi, cashFlows);
  if (npvLo * npvHi > 0) return null;
  for (let i = 0; i < maxIter; i++) {
    const mid = (lo + hi) / 2;
    const npvMid = npv(mid, cashFlows);
    if (Math.abs(npvMid) < tol) return mid;
    if (npvLo * npvMid <= 0) {
      hi = mid;
      npvHi = npvMid;
    } else {
      lo = mid;
      npvLo = npvMid;
    }
  }
  return (lo + hi) / 2;
}

export function breakevenLandValue(
  stabilizedNoi: number,
  targetCapRate: number,
): number | null {
  if (targetCapRate <= 0) return null;
  return stabilizedNoi / targetCapRate;
}

export type AgScenarioInput = {
  cashRentPerAcre: number;
  acres: number;
  vacancyRate: number;
  opexPerAcre: number;
  taxes: number;
  insurance: number;
  management: number;
  purchasePrice: number;
  holdYears: number;
  exitCapRate: number;
  annualAppreciation: number;
  discountRate: number;
};

export type AgScenarioResult = {
  grossRent: number;
  noi: number;
  capRate: number | null;
  cashOnCash: number | null;
  irr: number | null;
  npv: number;
  breakevenLandValue: number | null;
};

export function farmlandScenario(input: AgScenarioInput): AgScenarioResult {
  const grossRent = input.cashRentPerAcre * input.acres;
  const noi = noiFromRent(
    grossRent,
    input.vacancyRate,
    input.opexPerAcre * input.acres,
    input.taxes,
    input.insurance,
    input.management,
  );
  const flows: number[] = [-input.purchasePrice];
  for (let y = 1; y <= input.holdYears; y++) {
    flows.push(noi);
  }
  const exitValue =
    input.purchasePrice * Math.pow(1 + input.annualAppreciation, input.holdYears);
  // Approximate exit via appreciation path; also expose cap-based breakeven separately
  flows[flows.length - 1] += exitValue;
  return {
    grossRent,
    noi,
    capRate: capRate(noi, input.purchasePrice),
    cashOnCash: cashOnCash(noi, input.purchasePrice),
    irr: irr(flows),
    npv: npv(input.discountRate, flows),
    breakevenLandValue: breakevenLandValue(noi, input.exitCapRate),
  };
}
